File: providers/binance.go
package providers

import (
	"context"
	"encoding/json"
	"strconv"
	"sync"
	"time"

	"crypto-tray/internal/httpclient"
)

// Binance implements the Provider interface for the Binance API
type Binance struct {
	httpClient *httpclient.Client

	// Symbol cache
	cacheMu   sync.RWMutex
	symbols   []SymbolInfo
	coinIDMap map[string]SymbolInfo
	cacheTime time.Time
}

// NewBinance creates a new Binance provider
func NewBinance() *Binance {
	return &Binance{
		httpClient: httpclient.New(httpclient.Config{
			BaseURL: binanceBaseURL,
			Timeout: binanceTimeout,
		}),
	}
}

func (b *Binance) ID() string               { return "binance" }
func (b *Binance) Name() string             { return "Binance" }
func (b *Binance) RequiresAPIKey() bool     { return false }
func (b *Binance) SetAPIKey(key string)     {}
func (b *Binance) DefaultCoinIDs() []string { return []string{"BTCUSDT", "ETHUSDT", "SOLUSDT"} }

// FetchPrices retrieves prices for multiple cryptocurrencies by coinID (trading pair)
func (b *Binance) FetchPrices(ctx context.Context, coinIDs []string) ([]*PriceData, error) {
	if len(coinIDs) == 0 {
		return []*PriceData{}, nil
	}

	// Build JSON array format for symbols query parameter
	symbolsJSON, err := json.Marshal(coinIDs)
	if err != nil {
		return nil, err
	}

	query := BinanceTickerQuery{
		Symbols: string(symbolsJSON),
	}

	var tickers []BinanceTicker24hr
	if err := b.httpClient.GetWithQuery(ctx, "/api/v3/ticker/24hr", query, &tickers); err != nil {
		return nil, err
	}

	// Build response map for quick lookup
	tickerMap := make(map[string]BinanceTicker24hr)
	for tickerIdx := range tickers {
		t := tickers[tickerIdx]
		tickerMap[t.Symbol] = t
	}

	prices := make([]*PriceData, 0, len(coinIDs))
	for coinIdx := range coinIDs {
		coinID := coinIDs[coinIdx]
		ticker, ok := tickerMap[coinID]
		if !ok {
			continue
		}

		price, err := strconv.ParseFloat(ticker.LastPrice, 64)
		if err != nil {
			continue
		}

		change, err := strconv.ParseFloat(ticker.PriceChangePercent, 64)
		if err != nil {
			change = 0
		}

		prices = append(prices, &PriceData{
			CoinID:    coinID,
			Symbol:    b.coinIDToSymbol(coinID),
			Price:     price,
			Change24h: change,
		})
	}

	return prices, nil
}

File: src-tauri/src/providers/binance.rs
use async_trait::async_trait;
use reqwest::Client;
use serde::Deserialize;

use crate::error::Result;
use crate::providers::traits::Provider;
use crate::providers::types::{PriceData, ProviderInfo, SymbolInfo};

const BASE_URL: &str = "https://api.binance.com";

#[derive(Debug, Deserialize)]
struct TickerResponse {
    symbol: String,
    #[serde(rename = "lastPrice")]
    last_price: String,
    #[serde(rename = "priceChangePercent")]
    price_change_percent: String,
}

#[derive(Debug, Deserialize)]
struct ExchangeInfoResponse {
    symbols: Vec<ExchangeSymbol>,
}

#[derive(Debug, Deserialize)]
struct ExchangeSymbol {
    symbol: String,
    #[serde(rename = "baseAsset")]
    base_asset: String,
    #[serde(rename = "quoteAsset")]
    quote_asset: String,
    status: String,
}

pub struct BinanceProvider {
    client: Client,
}

impl BinanceProvider {
    pub fn new() -> Self {
        Self {
            client: Client::new(),
        }
    }
}

#[async_trait]
impl Provider for BinanceProvider {
    fn info(&self) -> ProviderInfo {
        ProviderInfo {
            id: "binance".to_string(),
            name: "Binance".to_string(),
            requires_api_key: false,
        }
    }

    fn set_api_key(&mut self, _key: String) {
        // Binance doesn't require API key for public endpoints
    }

    fn default_coin_ids(&self) -> Vec<String> {
        vec![
            "BTCUSDT".to_string(),
            "ETHUSDT".to_string(),
            "SOLUSDT".to_string(),
        ]
    }

    async fn fetch_symbols(&self) -> Result<Vec<SymbolInfo>> {
        let url = format!("{}/api/v3/exchangeInfo", BASE_URL);
        let response: ExchangeInfoResponse = self
            .client
            .get(&url)
            .send()
            .await?
            .json()
            .await?;

        let symbols = response
            .symbols
            .into_iter()
            .filter(|s| s.status == "TRADING" && s.quote_asset == "USDT")
            .map(|s| SymbolInfo {
                id: s.symbol,
                symbol: s.base_asset.clone(),
                name: s.base_asset,
            })
            .collect();

        Ok(symbols)
    }

    async fn fetch_prices(&self, symbols: &[String]) -> Result<Vec<PriceData>> {
        let url = format!("{}/api/v3/ticker/24hr", BASE_URL);
        let response: Vec<TickerResponse> = self
            .client
            .get(&url)
            .send()
            .await?
            .json()
            .await?;

        let prices = response
            .into_iter()
            .filter(|t| symbols.contains(&t.symbol))
            .map(|t| {
                let price = t.last_price.parse::<f64>().unwrap_or(0.0);
                let change = t.price_change_percent.parse::<f64>().unwrap_or(0.0);
                // Convert coinID (e.g., "BTCUSDT") to display symbol (e.g., "BTC")
                let display_symbol = t.symbol.strip_suffix("USDT").unwrap_or(&t.symbol).to_string();
                PriceData {
                    coin_id: t.symbol.clone(),
                    symbol: display_symbol,
                    name: String::new(),
                    price,
                    change_24h: change,
                    converted_price: 0.0, // Binance returns USD prices directly
                    currency: "usd".to_string(),
                }
            })
            .collect();

        Ok(prices)
    }
}
